pyquil/api/_devices.py

```diff
@@ -13,38 +13,79 @@
-        "noise_model": a NoiseModel object describing the entire device, serialized as a dictionary
-    }
+    NOTE: These can't directly be used to manufacture pyQuil Device objects, but this gives a list
+          of legal values that can be supplied to list_lattices to filter its (potentially very
+          noisy) output.
+
+    :return: A list of device names.
     """
+    # For the record, the dictionary stored in "devices" that we're getting back is keyed on device
+    # names and has this structure in its values:
+    #
+    # {
+    #   "is_online":   a boolean indicating the availability of the device,
+    #   "is_retuning": a boolean indicating whether the device is busy retuning,
+    #   "specs":       a Specs object describing the entire device, serialized as a dictionary,
+    #   "isa":         an ISA object describing the entire device, serialized as a dictionary,
+    #   "noise_model": a NoiseModel object describing the entire device, serialized as a dictionary
+    # }
+
     session = get_session()
     config = PyquilConfig()
 
-    return get_json(session, config.forest_url + "/devices")["devices"]
+    return sorted(get_json(session, config.forest_url + "/devices")["devices"].keys())
 
 
-def get_device(name: str):
+def list_lattices(device_name: str = None, num_qubits: int = None,
+                  connection: ForestConnection = None):
     """
-    Construct a Device object from a QCS-available device.
+    Query the Forest 2.0 server for its knowledge of lattices.  Optionally filters by underlying
+    device name and lattice qubit count.
 
-    :param name: Name of the desired device.
-    :return: A Device object.
+    :return: A dictionary keyed on lattice names and valued in dictionaries of the form
+             {
+               "device_name": device_name,
+               "qubits": num_qubits
+             }
     """
-    raise NotImplementedError('QPU devices will be available at a later date, please use the QVM')
+    if connection and connection.session:
+        session = connection.session
+    else:
+        session = get_session()
+
+    if connection:
+        url = connection.sync_endpoint + "/lattices"
+    else:
+        config = PyquilConfig()
+        url = config.forest_url + "/lattices",
+
+    try:
+        response = get_json(session, url,
+                            params={"device_name": device_name,
+                                    "num_qubits": num_qubits})
+
+        return response["lattices"]
+    except Exception as e:
+        raise ValueError("""
+        list_lattices encountered an error when querying the Forest 2.0 endpoint.
+
+        Very likely this means that you do not have valid user authentication information, and very
+        likely this is because you haven't yet been invited to try QCS.  We plan on making our
+        device information publicly accessible soon, but in the meanwhile, you'll have to use
```

Comment:
Another way this could happen, is if they have been invited and have all the necessary auth, but somehow deleted or made some unfortunate edits to their `.qcs_config/.forest_config`. I, and other internal users, have gotten to this spot by missing `config.forest_url`. It's a pretty awesome error message, but doesn't cover all the bases that I've seen. (Would particularly like this addressed so I can remove my other PR :innocent: )

Reply:
@lcapelluto  can you submit a "needs changes" review feedback so it doesn't look like this is ready to merge

Reply:
I pushed an expanded error message. Let me know if there's more to do.